lib/puma/server.rb

```diff
@@ -583,32 +583,35 @@ def response_to_error(client, requests, err, status_code)
     # Wait for all outstanding requests to finish.
     #
     def graceful_shutdown
-      if options[:shutdown_debug]
-        threads = Thread.list
-        total = threads.size
+      shutdown_debugger = if options[:shutdown_debug]
+        lambda do |message|
+          threads = Thread.list
+          total = threads.size
 
-        pid = Process.pid
+          pid = Process.pid
 
-        $stdout.syswrite "#{pid}: === Begin thread backtrace dump ===\n"
+          $stdout.syswrite "#{pid}: #{message}\n"
+          $stdout.syswrite "#{pid}: === Begin thread backtrace dump ===\n"
 
-        threads.each_with_index do |t,i|
-          $stdout.syswrite "#{pid}: Thread #{i+1}/#{total}: #{t.inspect}\n"
-          $stdout.syswrite "#{pid}: #{t.backtrace.join("\n#{pid}: ")}\n\n"
+          threads.each_with_index do |t,i|
+            $stdout.syswrite "#{pid}: Thread #{i+1}/#{total}: #{t.inspect}\n"
+            $stdout.syswrite "#{pid}: #{t.backtrace.join("\n#{pid}: ")}\n\n"
+          end
+
+          $stdout.syswrite "#{pid}: === End thread backtrace dump ===\n"
         end
-        $stdout.syswrite "#{pid}: === End thread backtrace dump ===\n"
+      end
+
+      if shutdown_debugger && options[:shutdown_debug] != :on_force
+        shutdown_debugger.call("Shutdown initiated")
+        shutdown_debugger = nil
       end
 
       if @status != :restart
         @binder.close
       end
 
-      if @thread_pool
-        if timeout = options[:force_shutdown_after]
-          @thread_pool.shutdown timeout.to_f
-        else
-          @thread_pool.shutdown
-        end
-      end
+      @thread_pool.shutdown options[:force_shutdown_after], shutdown_debugger
```

Comment:
1. Reading the code I'm quite sure that `@thread_pool` will always be truthy at this point, so we don't need the first `if`.
2. I've moved the default `:force_shutdown_after` value (-1) to the config  so we no longer need to worry about passing in `nil` and overriding the default in the method definition i.e., we don't need the second `if` (and `else`).